File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/CPP/ReverseLinkedListII.cpp
/*  
  Reverse a linked list from position m to n. Do it in-place and in one-pass.
  
  For example:
  Given 1->2->3->4->5->NULL, m = 2 and n = 4,
  
  return 1->4->3->2->5->NULL.
  
  Note:
  Given m, n satisfy the following condition:
  1 ? m ? n ? length of list.
*/


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *reverseBetween(ListNode *head, int m, int n) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if( m >= n || m <=0 )
            return head;
            
        int size=0;
        ListNode* temp=head;
        for(  ; temp != NULL; temp=temp->next , size ++ ); 
        if( size < n )
            return head;
        
        temp = head;
        ListNode* prev = NULL;
        int times = n - m;
        
        if( m == 1 )
            head = reverseList( prev, temp, times);
        else{
            for( ; (m--) > 1; prev = temp, temp = temp->next ) ;
            temp = reverseList( prev, temp, times );
        }
        return head;
    }
    
    ListNode* reverseList( ListNode* prev, ListNode* head,int times ){
        
        ListNode* temp = head;
        vector< ListNode* > nodes;
        for( int i=0; i<=times; i++){
            nodes.push_back( temp );
            temp = temp -> next;
        }
        
        nodes[0]->next = nodes[ times ] -> next;
        for( int i =times; i >0; i-- ){
            nodes[i]->next = nodes[i-1];
        }
        
        if( prev == NULL)
            return nodes[times];
        
        prev ->next = nodes[times];
        
        return nodes[times - 1];
        
    }
};

File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Python/ReverseLinkedListII.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

# For example:
# Given 1->2->3->4->5->NULL, m = 2 and n = 4,
#
# return 1->4->3->2->5->NULL. 


class Solution:
    # @param head, a ListNode
    # @param m, an integer
    # @param n, an integer
    # @return a ListNode
    def reverseBetween(self, head, m, n):
        """ reverse a linked list from m to n, do it in one pass"""
        if m == 1:
            return self.reverseFrom( head, n - m + 1) 
        else:
            # get the mth node
            prev, curr = None, head
            for _ in range(m - 1):
                if curr:
                    prev, curr = curr, curr.next 
                 
            prev.next = self.reverseFrom( curr, n - m + 1) 
            
            # since m > 1, head doesn't change
            return head

    def reverseFrom(self, head, length):
        """ 
        Reverse a linked list from the head to length th node
        Let's assume length >= len(head_list)
        Return a new head
        """
        if not head:
            return head

        sentry = ListNode(0)
        sentry.next = head

        prev, curr = None, head
        #for _ in range(length):
        while curr and length != 0:
            length = length - 1
            next_node = curr.next
            curr.next = prev
            
            prev = curr
            curr = next_node
        # previous head's next node is curr
        sentry.next.next = curr
        # prev is the new header
        return prev
            
def printList(node):
    #print "test"
    while node:
        print node.val
        node = node.next
#if __name__ == "__main__":
#    print "test"            
if __name__ == "__main__":
    node_1 = ListNode(10)
    node_2 = ListNode(20)
    node_3 = ListNode(30)
    node_4 = ListNode(40)
    node_5 = ListNode(50)

    node_1.next = node_2
    node_2.next = node_3
    node_3.next = node_4
    node_4.next = node_5
    printList(node_1) 
